types/rest/rest.go

```
@@ -197,6 +197,22 @@ func ParseFloat64OrReturnBadRequest(w http.ResponseWriter, s string, defaultIfEm
 	return n, true
 }
 
+// ParseQueryHeightOrReturnBadRequest sets the height to execute a query if set by the http request.
+func ParseQueryHeightOrReturnBadRequest(w http.ResponseWriter, cliCtx context.CLIContext, r *http.Request) (context.CLIContext, bool) {
+	heightStr := r.FormValue("height")
+	if heightStr != "" {
+		height, err := strconv.Atoi(heightStr)
```

What happens when `height < 0`?

---

What if height = 0 ?

---

(Idk if it’s a valid block number)

---

When `height = 0` data for the latest committed block is returned (https://tendermint.com/docs/app-dev/abci-spec.html#request-response-messages). Validation should fail early if a non-positive value is passed.

---

Correct, providing 0 is equivalent to not providing a height at all.

---

> Validation should fail early if a non-positive value is passed.

This still stands. Passing `height < 0` leads to undefined behavior as per specs. We should validate user's input more strictly here.

---

It will just default to the latest height (i.e. as if you passed in zero).